compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1248,11 +1267,22 @@ impl<'tcx> Predicate<'tcx> {
         // substitution code expects equal binding levels in the values
         // from the substitution and the value being substituted into, and
         // this trick achieves that).
-        let substs = trait_ref.skip_binder().substs;
-        let pred = self.skip_binders();
+        let bound_pred = self.bound_atom();
+        let trait_bound_vars = trait_ref.bound_vars();
+        let substs = tcx
+            .shift_bound_var_indices(bound_pred.bound_vars().len(), trait_ref.skip_binder())
+            .substs;
+        let pred = bound_pred.skip_binder();
         let new = pred.subst(tcx, substs);
         if new != pred {
-            ty::Binder::bind(new).potentially_quantified(tcx, PredicateKind::ForAll)
+            let bound_vars =
+                tcx.mk_bound_variable_kinds(bound_pred.bound_vars().iter().chain(trait_bound_vars));
+            if poly {
+                PredicateKind::ForAll(ty::Binder::bind_with_vars(new, bound_vars)).to_predicate(tcx)
+            } else {
```

Comment:
I think we just switch to always using `ForAll`.

IIRC, we only need to keep bound vars because a projection was referencing a late bound var that didn't appear in the input types (which is a semantic error but we still gracefully handle it)